### ISSUE TITLE
Threshold to predict new data

### ISSUE THREAD

**Issue by Hari-pyt** [bradleypallen/keras-quora-question-pairs#5]
can i include round() in the final layer

or just put a threshold 0.5 for the sigmoid outputs for new predictions

**Comment by marcoaleixo** [comment]
how did you compiled the model? I'm begineer to :D

**Comment by Hari7696** [comment]
Compiled the model in my local windows machine

python quora_question_pairs.py (didnt used jupyter)

each epoch takes 360 seconds aprox, cpu tensrflow
takes 2.5 hrs for training

output pic

![image](https://user-images.githubusercontent.com/32614314/38105985-45f96a2e-33ab-11e8-954c-4505736c42ec.png)

#1 customised your code for prediction
predict.py

print("Processing", QUESTION_PAIRS_FILE)
question1 = []
question2 = []
   
with open(KERAS_DATASETS_DIR + QUESTION_PAIRS_FILE, encoding='utf-8') as csvfile:
    reader = csv.DictReader(csvfile, delimiter=',')
    for row in reader:
        question1.append(row['question1'])
        question2.append(row['question2'])

print('Question pairs: %d' % len(question1))
questions = question1 + question2
tokenizer = Tokenizer(num_words=MAX_NB_WORDS)
tokenizer.fit_on_texts(questions)
question1_word_sequences = tokenizer.texts_to_sequences(question1)
question2_word_sequences = tokenizer.texts_to_sequences(question2)
word_index = tokenizer.word_index

print("Words in index: %d" % len(word_index))

q1_data = pad_sequences(question1_word_sequences, maxlen=MAX_SEQUENCE_LENGTH)
q2_data = pad_sequences(question2_word_sequences, maxlen=MAX_SEQUENCE_LENGTH)

print('Shape of question1 data tensor:', q1_data.shape)
print('Shape of question2 data tensor:', q2_data.shape)
question1 = Input(shape=(MAX_SEQUENCE_LENGTH,))
question2 = Input(shape=(MAX_SEQUENCE_LENGTH,))
q1 = Embedding(nb_words + 1, 
                 EMBEDDING_DIM, 
                 weights=[word_embedding_matrix], 
                 input_length=MAX_SEQUENCE_LENGTH, 
                 trainable=False)(question1)
q1 = TimeDistributed(Dense(EMBEDDING_DIM, activation='relu'))(q1)
q1 = Lambda(lambda x: K.max(x, axis=1), output_shape=(EMBEDDING_DIM, ))(q1)
q2 = Embedding(nb_words + 1, 
                 EMBEDDING_DIM, 
                 weights=[word_embedding_matrix], 
                 input_length=MAX_SEQUENCE_LENGTH, 
                 trainable=False)(question2)
q2 = TimeDistributed(Dense(EMBEDDING_DIM, activation='relu'))(q2)
q2 = Lambda(lambda x: K.max(x, axis=1), output_shape=(EMBEDDING_DIM, ))(q2)

merged = concatenate([q1,q2])
merged = Dense(200, activation='relu')(merged)
merged = Dropout(DROPOUT)(merged)
merged = BatchNormalization()(merged)
merged = Dense(200, activation='relu')(merged)
merged = Dropout(DROPOUT)(merged)
merged = BatchNormalization()(merged)
merged = Dense(200, activation='relu')(merged)
merged = Dropout(DROPOUT)(merged)
merged = BatchNormalization()(merged)
merged = Dense(200, activation='relu')(merged)
merged = Dropout(DROPOUT)(merged)
merged = BatchNormalization()(merged)
is_duplicate = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[question1,question2], outputs=is_duplicate)
model.load_weights(MODEL_WEIGHTS_FILE)

temp= model.predict([q1_data, q2_data])
df = pd.DataFrame(temp)
df.to_csv("output.csv")


#2 OUTPUT.CSV file
.csv file pic
![image](https://user-images.githubusercontent.com/32614314/38106240-fea48356-33ab-11e8-9474-79a8d1124f88.png)

But I need boolean values here
so i came up with this
temp= model.predict([q1_data, q2_data])
for i in range(0,T1):          # T1 is len(question1)
    if(temp[i]>=0.4093):    #0.4093 is taken as threshold, i know its not a proper one
        temp[i]=1
    else:
        temp[i]=0

df = pd.DataFrame(temp)
df.to_csv("output.csv")
![image](https://user-images.githubusercontent.com/32614314/38108268-e2b59b16-33b1-11e8-8c44-cd159f81bf39.png)


**can you plzz help me in choosing the threshold :)
is the threshold deals with loss function?**

**Comment by marcoaleixo** [comment]
how you fixed that?

**Comment by Hari-pyt** [comment]
Sigmoid units will always give the probability of which class it belongs. So in cross validation ( evaluation) it round up the probability value and got 81 percent accuracy.  So I placed a threshold on output file,  if probability>0.5 = 1 else 0.

**Comment by Hari-pyt** [comment]
Also learned that if we add round() in final layer, network struggle with no improvement in accuracy.
